### DISCUSSION TITLE
Best way to setup debug builds in the CI?

### DISCUSSION THREAD

**Discussion by dpc** [ipetkov/crane#140] [Q&A] [answered]
Right now `crane` default to `--release` builds, and rightly so - because that's kind of the default, where the artifacts built by the flake should be prod-usable.

However, when used in CI, it seems like a waste of time to build the release mode.

So here is the question - what would be the neatest way to parametrize the existing build (`flake.nix` with a complex setup like [I have](https://github.com/fedimint/fedimint/blob/da0dc1e5934e87853e68b3a30b7cc226ab4b02a4/flake.nix#L1) to expose debug-builds to use in CI.

**Comment by ipetkov** [comment, marked as answer]
Easiest way? Set `CARGO_PROFILE = "dev";` in the derivation and all cargo invocations will effectively run as `cargo --profile dev` (which is the default "debug" profile used by cargo)

It's worth noting that cargo considers debug and release artifacts to be different, so if your CI doesn't build production/release builds your consumers may need to build it themselves (but you might not care about this).

I'd also note that you should be careful not to mix debug `cargoArtifacts` built with a derivation which expects to do a release build. Cargo will happily compile everything, but effectively you won't get any benefits of artifact caching.

It's also technically possible to build and cache both release and debug artifacts, but that will result in a lot more data being cached/copied around, so pick your poison

```nix
# Build both debug and release artifacts, be careful with cache bloat!
let
  src = ./whatever;
  cargoArtifactsDebug = craneLib.buildDepsOnly {
    inherit src;
    CARGO_PROFILE = "dev";
  };

  cargoArtifactsReleaseAndDebug = craneLib.buildDepsOnly {
    inherit src;
    cargoArtifacts = cargoArtifactsDebug;
    CARGO_PROFILE = "release";
  };
in
{
  myCrateRelease = craneLib.buildPackage {
    inherit src;
    cargoArtifacts = cargoArtifactsReleaseAndDebug;
    CARGO_PROFILE = "release";
  };

  myCrateDebug = craneLib.buildPackage {
    inherit src;
    cargoArtifacts = cargoArtifactsReleaseAndDebug;
    CARGO_PROFILE = "dev";
    };
}
```

**Reply by dpc** [reply]
I was hoping to avoid duplicating all the derivations.

I guess I'll just make all the derivations a result of a lambda taking `CARGO_PROFILE` as an argument, and then duplicate everything in the `outputs`

**Reply by ipetkov** [reply]
You might not even need the lambda, just use `mapAttrs` with `.overrideAttrs (old: { CARGO_PROFILE = "dev"; })`

**Reply by dpc** [reply]
That's so good, maybe it should be illegal.

**Reply by dpc** [reply]
Oh, wait ... but this will not fix the CARGO_PROFILE in derivation pointed at by `cargoArtifacts`... Hmm...

**Reply by ipetkov** [reply]
D'oh, you're totally right! Guess the lambda it is then...

**Reply by ipetkov** [reply]
Actually I wonder if you could also recurse into `.overridrAttrs (old: { cargoArtifacts = old.cargoArtifacts.overrideAttrs (...); }` 🤔 

Probably not worth the headache to express with Nix but it might be doable

**Reply by dpc** [reply]
```nix
     let debug =
            let overrideCargoProfileRec = deriv: profile: deriv.overrideAttrs (oldAttrs: { CARGO_PROFILE = profile; cargoArtifacts = if oldAttrs ? "cargoArtifacts" && oldAttrs.cargoArtifacts != null then overrideCargoProfileRec oldAttrs.cargoArtifacts profile else null; });
            in builtins.mapAttrs (name: deriv: overrideCargoProfileRec deriv "dev") packages;
```

A little bit of recursion fixed it.
